### ISSUE TITLE
Stop the duplication of key ID conversions

### ISSUE THREAD

**Issue by hug-dev** [parallaxsecond/parsec#331]
In almost all of our providers we have the same methods: `create_key_id`, `insert_key_id`, `get_key_id` and `remove_key_id` that are wrappers around the `ManageKeyID` methods with as parameters, the specific ProviderID and the specific key ID type.

It should be possible to only have one copy of all of those methods using generic types (with trait bounds possibly).

**Comment by ionut-arm** [comment]
My idea for this is to create a `KeyInfoManagerClient` that will be instantiated per provider. The client would be responsible for handling generic functionality, whilst the providers would still have to perform the conversions or operations specific to their backend.

For example, the method for removing keys would have this signature:
```rust
impl KeyInfoManagerClient {
    pub fn remove_key(&self, key_triple: &KeyTriple) -> Result<()> { ... }
}
```

The functions for creating key IDs will still need to be implemented inside the provider modules because it is highly specific to the backends and would not gain much from abstracting. All other KIM methods (inserting, removing, and the various querying methods) would be wrapped by the client. The client is also responsible for handling the `Arc<RWLock<...>>`, removing that burden from the providers

**Comment by hug-dev** [comment]
I think I agree with the general idea, I have a few comments:

* to simplify, can those methods directly be implemented on `dyn ManageKeyInfo + Send + Sync` (as with `list_keys`) so that they can directly be used?
* instead of letting the providers handle the conversions, the method could take a generic parameter `T` that implements `Serialize` for `insert_key_info` and `Deserialize` for `get_key_info`. Inside those methods `bincode` can be used to perform the conversions.

So to summarize those kind of methods:
```rust
fn insert_key_info(&mut self, T: Serialize, KeyTriple, Attributes) -> Result<()>;
fn remove_key_info(&mut self, KeyTriple) -> Result<()>;
fn get_key_info(&self, KeyTriple) -> Result<(T: Deserialize, Attributes)>;
```

**Comment by ionut-arm** [comment]
> to simplify, can those methods directly be implemented on dyn ManageKeyInfo + Send + Sync (as with list_keys) so that they can directly be used?

Sure, I can add them to `ManageKeyInfo`, but will probably keep the client wrapper around that - I'm not a fan of having to deal with `Arc<RWLock<dyn ManageKeyInfo + Send + Sync>>` in every provider :) 

> instead of letting the providers handle the conversions, the method could take a generic parameter ...

Yep! Current signature for `insert_key_info`:
```rust
pub fn insert_key_info<T: Serialize>(
        &self,
        key_triple: KeyTriple,
        key_id: &T,
        attributes: Attributes,
    ) -> parsec_interface::requests::Result<()>
```

**Comment by hug-dev** [comment]
> Sure, I can add them to ManageKeyInfo, but will probably keep the client wrapper around that - I'm not a fan of having to deal with Arc<RWLock<dyn ManageKeyInfo + Send + Sync>> in every provider :)

Happy to use a wrapper if it's too complicated! From my experience with `list_keys` and `list_clients` though, with the `Deref` implemented all the way, there isn't much boilerplate

**Comment by ionut-arm** [comment]
Interesting question:

Should `remove_key_info` and `insert_key_info` fail if the key doesn't exist (for `remove`) or already exists (for `insert`)?

**Comment by hug-dev** [comment]
Yes I think they should! With like `PsaDoesNotExist` or `PsaAlreadyExist`, I think we do not have the same behaviour on all providers currently

**Comment by ionut-arm** [comment]
Done in #343
